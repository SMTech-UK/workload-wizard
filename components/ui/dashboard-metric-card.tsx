import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-500">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-muted-foreground mt-1">{subtitle}</div>}
    </div>
  );
} 