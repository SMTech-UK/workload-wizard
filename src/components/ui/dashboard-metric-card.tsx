import { ReactNode } from "react";

interface DashboardMetricCardProps {
  title: string;
  value: ReactNode;
  subtitle?: string;
  icon?: ReactNode;
  valueClassName?: string;
  children?: ReactNode; // For custom content like Progress bar
}

export function DashboardMetricCard({
  title,
  value,
  subtitle,
  icon,
  valueClassName = "",
  children,
}: DashboardMetricCardProps) {
  return (
    <div className="rounded-lg border p-6 bg-white dark:bg-zinc-900 shadow-sm">
      <div className="flex flex-row items-center justify-between space-y-0 pb-2">
        <span className="text-sm font-medium text-gray-900 dark:text-white">{title}</span>
        {icon}
      </div>
      <div className={`text-2xl font-bold ${valueClassName}`}>{value}</div>
      {children}
      {subtitle && <div className="text-xs text-gray-600 dark:text-gray-300 mt-1">{subtitle}</div>}
    </div>
  );
} 