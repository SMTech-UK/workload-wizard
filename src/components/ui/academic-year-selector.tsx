"use client"

import React from "react";
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from "@/components/ui/select";
import { Badge } from "@/components/ui/badge";
import { useAcademicYear } from "@/hooks/useAcademicYear";
import { Calendar } from "lucide-react";

interface AcademicYearSelectorProps {
  className?: string;
  showActiveBadge?: boolean;
  onAcademicYearChange?: (academicYearId: string) => void;
}

export function AcademicYearSelector({ 
  className, 
  showActiveBadge = true,
  onAcademicYearChange 
}: AcademicYearSelectorProps) {
  const { 
    academicYears, 
    currentAcademicYear, 
    currentAcademicYearId, 
    changeAcademicYear 
  } = useAcademicYear();

  const handleAcademicYearChange = async (academicYearId: string) => {
    try {
      await changeAcademicYear(academicYearId);
      onAcademicYearChange?.(academicYearId);
    } catch (error) {
      console.error('Failed to change academic year:', error);
    }
  };

  // Function to format academic year name (e.g., "2025/26" -> "AY-25/26")
  const formatAcademicYearName = (yearName: string) => {
    // Extract the last 2 digits from each year part
    const parts = yearName.split('/');
    if (parts.length === 2) {
      const startYear = parts[0].slice(-2);
      const endYear = parts[1].slice(-2);
      return `AY-${startYear}/${endYear}`;
    }
    // Fallback to original name if format is unexpected
    return yearName;
  };

  if (!academicYears || academicYears.length === 0) {
    return (
      <div className={`flex items-center gap-2 text-muted-foreground ${className}`}>
        <Calendar className="h-4 w-4" />
        <span className="text-sm">No academic years configured</span>
      </div>
    );
  }

  return (
    <div className={`flex items-center gap-2 ${className}`}>
      <Select
        value={currentAcademicYearId || ""}
        onValueChange={handleAcademicYearChange}
      >
        <SelectTrigger className="w-auto min-w-[120px] flex items-center gap-2">
          <Calendar className="h-4 w-4 text-muted-foreground" />
          <span className="font-medium">
            {currentAcademicYear ? formatAcademicYearName(currentAcademicYear.name) : "Select AY"}
          </span>
        </SelectTrigger>
        <SelectContent>
          {academicYears.map((year: any) => (
            <SelectItem key={year._id} value={year._id}>
              <div className="flex items-center gap-2">
                <Calendar className="h-4 w-4 text-muted-foreground" />
                <span>{formatAcademicYearName(year.name)}</span>
                {year.isActive && showActiveBadge && (
                  <Badge variant="default" className="text-xs">Active</Badge>
                )}
                {year.isStaging && (
                  <Badge variant="secondary" className="text-xs">Staging</Badge>
                )}
              </div>
            </SelectItem>
          ))}
        </SelectContent>
      </Select>
    </div>
  );
} 