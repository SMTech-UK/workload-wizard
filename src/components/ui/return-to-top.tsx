"use client";

import { Button } from "@/components/ui/button";
import { ChevronUp, Home } from "lucide-react";
import { useState, useEffect } from "react";
import Link from "next/link";

export function ReturnToTop() {
  const [isVisible, setIsVisible] = useState(true); // Temporarily always visible for testing

  useEffect(() => {
    const toggleVisibility = () => {
      console.log("Scroll position:", window.pageYOffset);
      if (window.pageYOffset > 300) {
        console.log("Setting visible to true");
        setIsVisible(true);
      } else {
        console.log("Setting visible to false");
        setIsVisible(false);
      }
    };

    // Call once on mount to check initial state
    toggleVisibility();
    
    window.addEventListener("scroll", toggleVisibility);
    return () => window.removeEventListener("scroll", toggleVisibility);
  }, []);

  const scrollToTop = () => {
    console.log("Scrolling to top");
    window.scrollTo({
      top: 0,
      behavior: "smooth",
    });
  };

  console.log("ReturnToTop component render, isVisible:", isVisible);

  return (
    <>
      {/* Return to Home button - top left */}
      <Link href="/">
        <Button
          size="sm"
          variant="outline"
          className="fixed top-4 left-4 z-50 shadow-lg hover:shadow-xl transition-all duration-200"
          aria-label="Return to home"
        >
          <Home className="h-4 w-4 mr-2" />
          Home
        </Button>
      </Link>

      {/* Return to Top button - bottom left */}
      <Button
        onClick={scrollToTop}
        size="sm"
        className="fixed bottom-4 left-4 z-50 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 bg-blue-500 text-white"
        aria-label="Return to top"
      >
        <ChevronUp className="h-4 w-4" />
      </Button>
    </>
  );
} 