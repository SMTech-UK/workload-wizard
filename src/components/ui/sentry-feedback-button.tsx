"use client";

import { useEffect, useState } from "react";
import * as Sentry from "@sentry/nextjs";
import { Button } from "@/components/ui/button";
import { Bug, MessageCircle } from "lucide-react";
import { useModalAwareButton } from "@/hooks/useModalAwareButton";
import { createPortal } from "react-dom";

export function SentryFeedbackButton() {
  const [feedback, setFeedback] = useState<any>(null);
  const [widget, setWidget] = useState<any>(null);
  const [mounted, setMounted] = useState(false);
  const buttonRef = useModalAwareButton();

  useEffect(() => {
    setMounted(true);
    // Get the feedback instance on the client side
    setFeedback(Sentry.getFeedback());
  }, []);

  const handleFeedbackClick = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    
    if (!feedback) return;

    if (widget) {
      // If widget is open, close it
      widget.removeFromDom();
      setWidget(null);
    } else {
      // Create and show the form immediately
      feedback.createForm().then((newWidget: any) => {
        newWidget.appendToDom();
        newWidget.open();
        setWidget(newWidget);
      }).catch((error: any) => {
        console.error('Failed to create Sentry feedback form:', error);
      });
    }
  };

  if (!feedback || !mounted) return null;

  const button = (
    <Button
      ref={buttonRef}
      variant="destructive"
      size="lg"
      onClick={handleFeedbackClick}
      onMouseDown={(e) => {
        e.preventDefault();
        e.stopPropagation();
      }}
      className="fixed bottom-4 right-4 z-[99999] shadow-lg"
      style={{
        zIndex: 99999,
        pointerEvents: 'auto',
      }}
      aria-label="Report a bug or give feedback"
    >
      <Bug className="h-4 w-4 mr-2" />
      Report an Issue
    </Button>
  );

  // Render the button in a portal to ensure it's outside modal context
  return createPortal(button, document.body);
} 