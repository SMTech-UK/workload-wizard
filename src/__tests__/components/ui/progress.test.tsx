import { render, screen } from '@testing-library/react';
import { Progress } from '@/components/ui/progress';

describe('Progress Component', () => {
  describe('Rendering', () => {
    it('should render progress bar with default value', () => {
      // Arrange
      const value = 50;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toBeInTheDocument();
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(progress).toHaveAttribute('aria-valuemin', '0');
      expect(progress).toHaveAttribute('aria-valuemax', '100');
      expect(container).toMatchSnapshot();
    });

    it('should render progress bar with custom value', () => {
      // Arrange
      const value = 75;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(container).toMatchSnapshot();
    });

    it('should render progress bar with zero value', () => {
      // Arrange
      const value = 0;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(container).toMatchSnapshot();
    });

    it('should render progress bar with maximum value', () => {
      // Arrange
      const value = 100;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(container).toMatchSnapshot();
    });

    it('should render progress bar with custom className', () => {
      // Arrange
      const value = 50;
      const customClass = 'custom-progress';
      
      // Act
      const { container } = render(<Progress value={value} className={customClass} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass(customClass);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Accessibility', () => {
    it('should have proper ARIA attributes', () => {
      // Arrange
      const value = 60;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(progress).toHaveAttribute('aria-valuemin', '0');
      expect(progress).toHaveAttribute('aria-valuemax', '100');
      expect(container).toMatchSnapshot();
    });

    it('should have proper role attribute', () => {
      // Arrange
      const value = 30;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Edge Cases', () => {
    it('should handle negative values gracefully', () => {
      // Arrange
      const value = -10;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(container).toMatchSnapshot();
    });

    it('should handle values over 100 gracefully', () => {
      // Arrange
      const value = 150;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveAttribute('aria-valuenow', value.toString());
      expect(container).toMatchSnapshot();
    });

    it('should handle undefined value gracefully', () => {
      // Arrange
      const value = undefined;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Styling', () => {
    it('should have proper base styling classes', () => {
      // Arrange
      const value = 50;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      expect(progress).toHaveClass('relative', 'h-2', 'w-full', 'overflow-hidden', 'rounded-full', 'bg-secondary');
      expect(container).toMatchSnapshot();
    });
  });

  describe('Component Structure', () => {
    it('should render progress indicator with correct styling', () => {
      // Arrange
      const value = 50;
      
      // Act
      const { container } = render(<Progress value={value} />);
      
      // Assert
      const progress = screen.getByRole('progressbar');
      const indicator = progress.querySelector('[data-radix-progress-indicator]');
      expect(indicator).toBeInTheDocument();
      expect(container).toMatchSnapshot();
    });
  });

  describe('Snapshot Testing', () => {
    it('should maintain consistent UI structure across different values', () => {
      // Arrange & Act
      const { container: container1 } = render(<Progress value={25} />);
      const { container: container2 } = render(<Progress value={50} />);
      const { container: container3 } = render(<Progress value={75} />);
      
      // Assert
      expect(container1).toMatchSnapshot('progress-25-percent');
      expect(container2).toMatchSnapshot('progress-50-percent');
      expect(container3).toMatchSnapshot('progress-75-percent');
    });

    it('should maintain consistent UI structure with custom styling', () => {
      // Arrange & Act
      const { container } = render(
        <Progress 
          value={60} 
          className="h-4 bg-blue-100 border-2 border-blue-300" 
        />
      );
      
      // Assert
      expect(container).toMatchSnapshot('progress-custom-styling');
    });
  });
}); 