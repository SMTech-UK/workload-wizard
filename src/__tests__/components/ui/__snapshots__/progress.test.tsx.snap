// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress Component Accessibility should have proper ARIA attributes 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="60"
    aria-valuetext="60%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="60"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="60"
      style="transform: translateX(-40%);"
    />
  </div>
</div>
`;

exports[`Progress Component Accessibility should have proper role attribute 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="30"
    aria-valuetext="30%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="30"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="30"
      style="transform: translateX(-70%);"
    />
  </div>
</div>
`;

exports[`Progress Component Edge Cases should handle undefined value gracefully 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="indeterminate"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="indeterminate"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress Component Rendering should render progress bar with custom className 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="50%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20 custom-progress"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component Rendering should render progress bar with custom value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="75"
    aria-valuetext="75%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="75"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="75"
      style="transform: translateX(-25%);"
    />
  </div>
</div>
`;

exports[`Progress Component Rendering should render progress bar with default value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="50%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component Rendering should render progress bar with maximum value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="100"
    aria-valuetext="100%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="complete"
    data-value="100"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="complete"
      data-value="100"
      style="transform: translateX(-0%);"
    />
  </div>
</div>
`;

exports[`Progress Component Rendering should render progress bar with zero value 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="0"
    aria-valuetext="0%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="0"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="0"
      style="transform: translateX(-100%);"
    />
  </div>
</div>
`;

exports[`Progress Component Snapshot Testing should maintain consistent UI structure across different values: progress-25-percent 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="25"
    aria-valuetext="25%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="25"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="25"
      style="transform: translateX(-75%);"
    />
  </div>
</div>
`;

exports[`Progress Component Snapshot Testing should maintain consistent UI structure across different values: progress-50-percent 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="50"
    aria-valuetext="50%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="50"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="50"
      style="transform: translateX(-50%);"
    />
  </div>
</div>
`;

exports[`Progress Component Snapshot Testing should maintain consistent UI structure across different values: progress-75-percent 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="75"
    aria-valuetext="75%"
    class="relative h-2 w-full overflow-hidden rounded-full bg-primary/20"
    data-max="100"
    data-state="loading"
    data-value="75"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="75"
      style="transform: translateX(-25%);"
    />
  </div>
</div>
`;

exports[`Progress Component Snapshot Testing should maintain consistent UI structure with custom styling: progress-custom-styling 1`] = `
<div>
  <div
    aria-valuemax="100"
    aria-valuemin="0"
    aria-valuenow="60"
    aria-valuetext="60%"
    class="relative w-full overflow-hidden rounded-full h-4 bg-blue-100 border-2 border-blue-300"
    data-max="100"
    data-state="loading"
    data-value="60"
    role="progressbar"
  >
    <div
      class="h-full w-full flex-1 bg-primary transition-all"
      data-max="100"
      data-state="loading"
      data-value="60"
      style="transform: translateX(-40%);"
    />
  </div>
</div>
`;
